tools/src/main/java/com/example/core/broker/testkit/TopicStub.java

```diff
@@ -0,0 +1,50 @@
+package com.example.core.broker.testkit;
+
+import akka.actor.ActorRef;
+import akka.actor.ActorSystem;
+import akka.actor.Props;
+import akka.stream.Materializer;
+import com.lightbend.lagom.javadsl.api.broker.Subscriber;
+import com.lightbend.lagom.javadsl.api.broker.Topic;
+
+/**
+ * Stubs a {@link Topic} to run tests without starting the complete broker infrastructure. While {@link TopicStub}
+ * should only be used inside the Stub's of your Service, you can distribute the {@link ProducerStub} returned by
+ * {@link TopicStub#producerStub()} so that clients (test code) can send messages via this stubbed topic.
+ *
+ * @param <T>
+ */
+public class TopicStub<T> implements Topic<T> {
+
+    private final Materializer materializer;
+    private final ActorRef bufferActor;
+    private String topicId;
+
+    TopicStub(String topicId, ActorSystem actorSystem, Materializer materializer) {
+        Props bufferProps = Props.create(TopicBufferActor.class);
+        bufferActor = actorSystem.actorOf(bufferProps);
+
+        this.materializer = materializer;
+        this.topicId = topicId;
+    }
+
+    @Override
+    public Topic.TopicId topicId() {
+        return TopicId.of(topicId);
+    }
+
+    @Override
+    public Subscriber<T> subscribe() {
+        return new SubscriberStub<>(topicId, bufferActor, materializer);
+    }
+
+    /**
+     * @return a {@link ProducerStub} that clients (test code) can use to send messages through this {@link TopicStub}.
+     */
+    public ProducerStub<T> producerStub() {
+        return (T message) -> {
+            bufferActor.tell(message, ActorRef.noSender());
+        };
```

Comment:
The code, as is, allows creating many producers for a single topic like you'd get in a real brokering environment. Having many producers for a single topic in test wouldn't make much  sense though, so I agree with you that my original motivation is gone.

In general, I am trying to avoid adding methods to an impl of `Topic`. I want `TopicStub` to be just an in-memory `Topic`.
Also, I don't like that this design has 3 public classes either (`ProducerStub`, `TopicStub` and `TopicFactoryStub`). I would like to reduce it to a minimum and I'm trying to hide `TopicStub` completely so the user only sees `ProducerStub` and a `SomethingFactory`:

```
        MyServiceStub(TopicStubFactory topicFactory) {
            myProducerStub = topicFactory.producer(MyEvent.class, TOPIC_ID);
        }
        @Override
        public Topic<MyEvent> messageTopic() {
            return myProducerStub.topic();
        }
```
and then use it for sending:
```
       myProducerStub.send(MyEvent.of("the message"));
```


Probably I should rename the `TopicStubFactory` to `ProducerFactory` or `TopicProducerTestkit` but there's an (unrelated) `TopicProducer` in production code and I don't want to mix things up.